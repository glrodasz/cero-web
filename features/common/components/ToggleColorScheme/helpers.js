export const persistColorScheme = ({ isDarkMode, setIsDarkMode }) => {
  const colorScheme = isDarkMode ? 'dark' : 'light'
  document.querySelector('body').dataset.colorScheme = colorScheme
  localStorage.setItem('prefers-color-scheme', colorScheme)
  setIsDarkMode(isDarkMode)
}

export const loadAndListenColorScheme = ({
  setIsDarkMode,
  __persistColorScheme = persistColorScheme,
}) => {
  const darkModeMediaQuery = window.matchMedia('(prefers-color-scheme: dark)')
  darkModeMediaQuery?.addListener((event) => {
    const isDarkMode = event.matches
    __persistColorScheme({ isDarkMode, setIsDarkMode })
  })

  const localStorageColorScheme = localStorage.getItem('prefers-color-scheme')

  if (localStorageColorScheme) {
    const isDarkMode = localStorageColorScheme === 'dark'
    __persistColorScheme({ isDarkMode, setIsDarkMode })
  } else {
    const isDarkMode = darkModeMediaQuery?.matches
    __persistColorScheme({ isDarkMode, setIsDarkMode })
  }
}
