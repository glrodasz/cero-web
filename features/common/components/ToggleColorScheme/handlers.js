import { persistColorScheme } from './helpers'

export const handleClick = ({ isDarkMode, setIsDarkMode }) => () => {
  persistColorScheme({ isDarkMode: !isDarkMode, setIsDarkMode })
}
