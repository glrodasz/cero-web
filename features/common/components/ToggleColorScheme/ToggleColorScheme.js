import { useState, useEffect } from 'react'
import { Check, Heading } from '@glrodasz/components'

import { handleClick } from './handlers'
import { loadAndListenColorScheme } from './helpers'

const ToggleColorScheme = () => {
  const [isDarkMode, setIsDarkMode] = useState(false)

  useEffect(() => {
    loadAndListenColorScheme({ setIsDarkMode })
  }, [])

  return (
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
        cursor: 'pointer',
        width: 150,
      }}
      onClick={handleClick({ isDarkMode, setIsDarkMode })}
    >
      <Check isChecked={isDarkMode} /> <Heading>Dark Mode</Heading>
    </div>
  )
}

export default ToggleColorScheme
