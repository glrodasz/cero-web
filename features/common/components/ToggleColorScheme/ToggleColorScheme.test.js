import ToggleColorScheme from './ToggleColorScheme'
import { render } from '@testing-library/react'

jest.mock('@glrodasz/components', () => ({
  Check: (props) => `${JSON.stringify(props)}`,
  Heading: ({ children }) => children,
}))

jest.mock('./helpers.js', () => ({
  loadAndListenColorScheme: () => {},
}))

describe.only('[ features / common / ToggleColorScheme ]', () => {
  describe('when ToggleColorScheme is called', () => {
    it('should render', () => {
      const { asFragment } = render(<ToggleColorScheme />)
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
