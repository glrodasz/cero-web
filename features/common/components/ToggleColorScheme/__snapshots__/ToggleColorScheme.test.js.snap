// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ features / common / components / ToggleColorScheme ] when \`ToggleColorScheme\` is mounted should render 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center; cursor: pointer; width: 150px;"
  >
    (Check {"isChecked":false} /)  (Heading {})Dark Mode(/Heading) 
  </div>
</DocumentFragment>
`;
